Select input improvements: Arrow navigation & text filtering

estib-vega opened the pull request:
Related to #2590

This adds type-to-filter and up-down-arrow navigation to the Select component. 
This should enable easy & fast switching between remote branches 😊

This is how it looks when using it on my local fork of gitbutler.
Once the dropdown is open, one can navigate up and down the options and click enter to select.
Added to that, one can type text (shown in the Select component's text field) and the list will be filtered based on whether or not the typed context is included in the label displayed.

![select-optimization](https://github.com/gitbutlerapp/gitbutler/assets/35891811/dc28bccf-642d-45c3-b58b-f10dd36d03dc)

### Style
The way this shows what option is being 'highlighted' when navigating with the arrows, if by changing the background to
```
background-color: var(--clr-bg-3);
```

I'm not sure this is consistent with design guidelines or not. Happy to change that accordingly if not

Caleb-T-Owens commented on app/src/lib/utils/string.ts:
```diff
@@ -11,3 +11,23 @@ export function hashCode(s: string) {
+ * @returns `true` if the string is a single character, `false` otherwise.
+ */
+export function isChar(char: string) {
+	return char.length === 1;
+}
+
+/**
+ * Checks if a value is a string.
+ *
+ * @param s - The value to check.
+ * @returns A boolean indicating whether the value is a string.
+ */
+export function isStr(s: unknown): s is string {
+	return typeof s === 'string';
```
Please use `==`

estib-vega replied:
Will do in a follow-up commit, but would you be please be open to reconsider this position at a later date?

Not only there are some cases in which even using TypeScript (e.g. when using union types) you need to use strict equality, but it's also just [faster than loose equality](https://stackoverflow.com/questions/12374815/javascript-vs-operators-performance).

Given, the performance gains are not huge but they are not 0. And I'd say that that in combination with the cases in which is still needed in TS it's something that should win against cross-language stylistic choices.

Ok that's what I needed to say haha 😅

Caleb-T-Owens replied:
Hey! I'm interested, could you share an example of how a union can be problematic?

estib-vega replied:
Gladly!
The simplest example would be a something like this:
```ts
let example: string | number[];

function initializeExample() {
  example = [1];
}

initializeExample();

if (example == "1") {
  console.log("unsafe");
} else {
  console.log("safe");
}

// Console will log: unsafe
```
The TS linter would not catch this because the value is being set outside of the immediate scope in which the `if` statement is being evaluated, so it will still treat the union type as `string | number[]` instead of `number[]`.

Another example is with `any` which just completely disables TypeScript checks.
```ts
let example: any;

function initializeExample() {
  example = [1,2];
}

initializeExample();

if (example == "1,2") {
  console.log("unsafe");
} else {
  console.log("safe");
}

// Console will log: unsafe
```

Caleb-T-Owens replied:
Hi @estib-vega,

We've talked about it internally, and we'll be moving over to `===`, so feel free to revert the commit, or I'll be coming in later this week to convert the project.

Caleb-T-Owens commented on app/src/lib/ai/types.ts:
```diff
@@ -39,11 +39,11 @@ export interface AIClient {
 
-export interface UserPrompt {
+export type UserPrompt = {
 	id: string;
 	name: string;
 	prompt: Prompt;
-}
+};
```
Is there a reason to prefer a type over an interface here?

estib-vega replied:
In this specific commit, just so that the shape can be inferred into the `SelectItemType extends Record<string, unknown>` generic. Interfaces can't be inferred as something that extends `Record`s

In general, just because `type` will raise compile errors if there are any other types in the same scope. While `interface`s will just merge both by default, which is not always what one wants.
It's not that common so it's not necessary a huge improvement.

estib-vega replied:
Reference 😊 https://github.com/microsoft/TypeScript/issues/15300#issuecomment-332366024

mtsgrd commented on app/src/lib/components/ProjectSwitcher.svelte:
```diff
@@ -9,11 +9,26 @@
+
+	let mappedProjects: ProjectRecord[] = [];
+
+	projectService.projects.subscribe((projectList) => {
+		// Map the projectList to fit the ProjectRecord type
+		mappedProjects = projectList.map((project) => {
+			return {
+				id: project.id,
+				title: project.title
+			};
+		});
+	});
```
Thanks for contributing! Since this will never unsubscribe, could we turn it into a derived store?

estib-vega replied:
Sure thing!
I'll create a follow-up commit later today

mtsgrd reviewed (commented):
Again, thanks so much for contributing! Could you have a look at my comments and let me know what you think?

estib-vega commented on app/src/lib/components/ProjectSwitcher.svelte:
```diff
@@ -14,17 +15,13 @@
-			};
-		});
-	});
+	// Create a derived store that maps the projects to a simpler structure
```
Sounds good 👍

estib-vega commented on app/src/lib/utils/misc.ts:
```diff
@@ -0,0 +1,24 @@
+/**
+ * Throttles the execution of a function.
+ * @param fn The function to be throttled.
+ * @param wait The time to wait between function invocations in milliseconds.
+ * @returns A throttled version of the original function.
```
> Type params are redundant with typescript
>

Do you mean the `<T extends (...args: any[]) => any>`? I would disagree.
In this case it is accomplishing 3 things:
1. Narrows down the `fn` argument to be any kind of function
2. Ensures that the return type is always the exact same type of function as the one passed in as an argument.
3. Allows us to keep type-guards when accessing the parameters or return types, dynamically (see lines 12 and 14). 

Is that what you meant, or were you referring to something else?

> and perhaps we can omit the comment too since it's a simple function?
>

Will remove that 👍

mtsgrd replied:
Sorry, no I meant the jsdoc, not the actual parameter!

estib-vega replied:
Ah gotcha :)

estib-vega commented on app/src/lib/components/Select.svelte:
```diff
@@ -52,6 +77,80 @@
+		filterText = filterText.slice(0, -1);
+	}, INPUT_THROTTLE_TIME);
+
+	function handleKeyDown(event: CustomEvent<KeyboardEvent>) {
```
Sure thing 👍

estib-vega commented on app/src/lib/components/Select.svelte:
```diff
@@ -27,8 +39,21 @@
+		: (filteredItems = filterStringByKey(items, labelId, filterText));
+
+	// Set highlighted item based on index
+	$: highlightIndex !== undefined
+		? (highlightedItem = filteredItems[highlightIndex])
+		: (highlightedItem = undefined);
```
Yeah, that way is actually cleaner, will change that 😄

estib-vega commented on app/src/lib/components/Select.svelte:
```diff
@@ -1,18 +1,30 @@
 
+	const INPUT_THROTTLE_TIME = 100;
+
+	type SelectItemKeyType = keyof SelectItemType;
```
> Since it's already a type, could we drop the `Type` suffix?
>
Totally 👍 

> But also, since it's a keyof something that extends Record<string, unknown>, I don't see the value of this type.
>
It's so that we can narrow down the `labelId` and `itemId` properties (see lines 26 and 27). This way, the component dynamically checks what the item type is, and based on that we set the acceptable types for the `labelId` and `itemId`.

What this does in practice is for the `AIPromptSelect.svelte` component is that if we pass the `allPrompts: UserPrompt[]` variable as the `items` property, the component will infer that the only allowed keys are `"id"`, `"name"` or `"prompt"`

```ts
export type UserPrompt = {
	id: string;
	name: string;
	prompt: Prompt;
};
```

Which results in the auto-completion in the IDE:
<img width="443" alt="auto-suggest" src="https://github.com/gitbutlerapp/gitbutler/assets/35891811/1a52737e-23f1-4db9-baeb-38d3fcdad42e">

And if the wrong key is used or if it's updated, a TS compiler error will catch that:
<img width="568" alt="type-guard" src="https://github.com/gitbutlerapp/gitbutler/assets/35891811/529bf50e-e33a-415c-8820-927b5a755431">

mtsgrd replied:
Thanks, you're right. That's awesome.

estib-vega commented:
@mtsgrd I've added another commit addressing the feedback given.
Let me know if some of the responses to the feedback still don't seem good enough arguments for keeping some things as they are. 
I'd be happy to change them if still needed!

mtsgrd commented on app/src/lib/components/Select.svelte:
```diff
@@ -27,8 +39,18 @@
+	let filteredItems: Selectable[] = items;
 
-	function handleItemClick(item: any) {
+	$: filterText === undefined
+		? (filteredItems = items)
+		: (filteredItems = filterStringByKey(items, labelId, filterText));
```
This one too pls :)

estib-vega replied:
Totally, yeah 👍

mtsgrd commented on app/src/lib/utils/filters.ts:
```diff
@@ -1,4 +1,28 @@
+/**
+ * Filters an array of objects based on a specific string value in a given key.
+ *
+ * @template T - The type of objects in the array.
+ * @template K - The key of the object to filter by.
+ * @param {T[]} arr - The array of objects to filter.
+ * @param {K} key - The key to filter by.
+ * @param {string} value - The string value to search for in the specified key.
+ * @returns {T[]} - The filtered array of objects.
```
If you agree with the other comment about jsdc then we can probably omit this too?

estib-vega replied:
Ah yeah, sorry should have thought about that one. Will take it out too

mtsgrd commented:
So not directly about this pull request, but I'm curious, would you know how to generically type the <Modal> component to give us types for the `let:item` directive?

estib-vega commented:
> So not directly about this pull request, but I'm curious, would you know how to generically type the component to give us types for the `let:item` directive?

It's a good question. I'm not sure, would need to try it out.
I'm still learning how "advanced" TS interacts with Svelte. Plus my understanding is that generics themselves are relatively new to Svelte components so they might still have a funky behavior.

My first guess would be to just add the type generic type in the prop declaration and see if it works as we expect.

Do you have a specific use-case in mind 👀?

mtsgrd commented on app/src/lib/utils/string.ts:
```diff
@@ -11,3 +11,17 @@ export function hashCode(s: string) {
 	return hash.toString();
 }
+
+/**
+ * Checks if a string is a single character.
+ */
```
Same line of thought as earlier, let's drop these comments since it feels redundant.

estib-vega replied:
Will do 👍

mtsgrd commented on app/src/lib/utils/filters.ts:
```diff
@@ -1,4 +1,21 @@
 export function unique(value: any, index: number, array: any[]) {
 	return array.indexOf(value) === index;
 }
+
+/**
+ * Filters an array of objects based on a specific string value in a given key.
+ */
+export function filterStringByKey<T, K extends keyof T>(arr: T[], key: K, value: string): T[] {
```
I would again drop the comment since the function name is sufficient to understand what's going on. But, could we still rename the function? It looks like it would be more appropriately called `filterObjectsByKey`?

Caleb-T-Owens replied:
At the same time, I'm curious of the value of having this function at all given that it can be rather clearly as:

```ts
declare const items: ({ foo: number, bar: string })[]
const needle = "asfd"

items.filter(({ bar }) => bar.includes(needle))
```

Caleb-T-Owens replied:
To me writing the filter statement is clearer than calling a function called `filterObjectsBySpecifyPropertyThatIncludesString(objects, "key", "asfd")`

estib-vega replied:
I don't have strong feelings about either approach :).

The main 2 reasons for which I opted for a separate dedicated function:
- I think inlining functional array methods into certain expressions can be actually a bit harder to read. In this case we'd write that into the ternary operator:

```ts
	$: filteredItems =
		filterText === undefined ? items : items.filter((it) => it[labelId].includes(filterText));
```
vs

```ts
$: filteredItems =
		filterText === undefined ? items : filterObjectsByKey(items, labelId, filterText);
```

- Also, just because `for`-loops are [generally more performant](https://medium.com/@maxheadway/exploring-performance-traditional-loops-vs-functional-methods-in-javascript-76dedc749375) than functional array methods (which kind of suck because they are more elegant in most cases). In the case that the list of e.g. target branches is huge, filtering the list will become more expensive and we should use as much performance optimizations as we can.

In practice that might not be the case most times so if you still prefer the functional method I'm ok with thah 👍

estib-vega replied:
Also, sorry if I keep pushing back on some of the comments. I appreciate both of you being so patient 😄😄

Caleb-T-Owens replied:
> Also, sorry if I keep pushing back on some of the comments. I appreciate both of you being so patient 😄😄

I'm convinced that having discussion between multiple developers always ends up producing a much better result. Its great that you're pushing back and helping us consider the problem from your perspective. Please don't stop 😂!!

---

> Also, just because for-loops are [generally more performant](https://medium.com/@maxheadway/exploring-performance-traditional-loops-vs-functional-methods-in-javascript-76dedc749375) than functional array methods (which kind of sucks because they are more elegant in most cases). In the case that the list of e.g. target branches is huge, filtering the list will become more expensive and we should use as much performance optimizations as we can.

This is undoubtably true, but in the vast majority of situations the difference in performance is simply not relevant, even on lists of 10,000 odd items.

I think our general consensus is that code clarity should take a priority over scraping every last nanosecond of performance.

> I think inlining functional array methods into certain expressions can be actually a bit harder to read. In this case we'd write that into the ternary operator:

I'm still struggling to be convinced about your `filterObjectsByKey` because the name doesn't tell me that you're doing a string `.includes` on it. If anything `filterObjectsByKey` tells me that you've got a list of arbitrarily shaped objects and you're filtering out the objects which have one particular key.

I do see that you're point about the expression getting quite dense, so how about we define a function local to the component and say:

```ts

function filterItems(items: Item[], filterText: string | undefined) {
  if (filterText) {
    return items.filter((it) => it[labelId].includes(filterText));
  } else {
    return items
  }
}

$: filteredItems = filterItems(items, filterText)
```

By keeping the definition local, we're also leaving a convenient place where would could pull in a fuzzy comparison operation in the future

estib-vega replied:
> I'm convinced that having discussion between multiple developers always ends up producing a much better result. Its great that you're pushing back and helping us consider the problem from your perspective. Please don't stop 😂!!
>
Haha I agree 😄

> By keeping the definition local, we're also leaving a convenient place where would could pull in a fuzzy comparison operation in the future
>
Seems like a good compromise. Will add it as you described :)

Caleb-T-Owens commented on app/src/lib/components/Select.svelte:
```diff
@@ -27,8 +38,28 @@
+	}
 
-	function handleItemClick(item: any) {
+	$: filteredItems = filterText === undefined ? items : filterItems(items, filterText);
```
filterItems handles the case for when the filterText is undefined so this can be reduced to the following:
```suggestion
$: filteredItems = filterItems(items, filterText);
```

estib-vega replied:
Oh right, nice catch!
